ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerImpl.scala

```diff
@@ -40,29 +37,63 @@ private[services] final class TrackerImpl(
   override def track(request: SubmitAndWaitRequest)(implicit
       ec: ExecutionContext,
       loggingContext: LoggingContext,
-  ): Future[Either[CompletionFailure, CompletionSuccess]] = {
+  ): Future[Either[TrackedCompletionFailure, CompletionSuccess]] = {
     logger.trace("Tracking command")
-    val promise = Promise[Either[CompletionFailure, CompletionSuccess]]()
-    submitNewRequest(request, promise)
+    submitNewRequest(request)
   }
 
   private def submitNewRequest(
-      request: SubmitAndWaitRequest,
-      promise: Promise[Either[CompletionFailure, CompletionSuccess]],
+      request: SubmitAndWaitRequest
   )(implicit
       ec: ExecutionContext
-  ): Future[Either[CompletionFailure, CompletionSuccess]] = {
+  ): Future[Either[TrackedCompletionFailure, CompletionSuccess]] = {
+    val trackedPromise = Promise[Either[CompletionFailure, CompletionSuccess]]()
     queue
       .offer(
         Ctx(
-          promise,
+          trackedPromise,
           SubmitRequest(request.commands),
         )
       )
-      .andThen {
-        HandleOfferResult.completePromise(promise)
+      .flatMap[Either[TrackedCompletionFailure, CompletionSuccess]] {
+        case QueueOfferResult.Enqueued =>
+          trackedPromise.future.map(
+            _.left.map(completionFailure => QueueCompletionFailure(completionFailure))
+          )
+        case QueueOfferResult.Failure(t) =>
+          failedQueueSubmission(
+            GrpcStatus.ABORTED
+              .withDescription(s"Failed to enqueue: ${t.getClass.getSimpleName}: ${t.getMessage}")
+              .withCause(t)
+          )
+        case QueueOfferResult.Dropped =>
+          failedQueueSubmission(
+            GrpcStatus.RESOURCE_EXHAUSTED
+              .withDescription("Ingress buffer is full")
+          )
+        case QueueOfferResult.QueueClosed =>
+          failedQueueSubmission(GrpcStatus.ABORTED.withDescription("Queue closed"))
       }
-    promise.future
+      .recoverWith(transformQueueSubmissionExceptions)
+  }
+
+  private def failedQueueSubmission(status: GrpcStatus) =
+    Future.successful(Left(QueueSubmitFailure(status)))
+
+  private def transformQueueSubmissionExceptions
+      : PartialFunction[Throwable, Future[Either[TrackedCompletionFailure, CompletionSuccess]]] = {
+    case i: IllegalStateException
+        if i.getMessage == "You have to wait for previous offer to be resolved to send another request" =>
```

Comment:
Not sure tbh, I only moved the code.
Though looking through the code that message is part of akka internals so I don't see any easy way of handling this.

Reply:
The [docs](https://doc.akka.io/api/akka/2.6/akka/stream/scaladsl/SourceQueueWithComplete.html) are (for a change) not super-clear but it looks like this can only happen in this case (and the code seems to confirm), so I don't think there's a need to check the exception message.

Reply:
Fair enough. Thanks for considering it!